### PR TITLE
Handle invalid user dictionaries more robustly 

### PR DESCRIPTION
**Pull request by johnsyweb**
#### Context

We've had some application errors (`NoMethodError: undefined method 'downcase' for nil:NilClass`), which happens in the l33t matcher: https://github.com/envato/zxcvbn-ruby/blob/07ef78b72bcbd309baf2c843b739b38ac14bd696/lib/zxcvbn/matching/l33t.rb#L32

Due to the way we conceal personally identifiable information, we cannot see the input values that are triggering the problem.

This PR is the result of me trying (and failing) to reproduce the error in this gem.

#### Change

 - Improved a bunch of tests
 - Added [Guard](https://guardgem.org/) gem for faster feedback
 - Handle invalid user dictionaries more robustly (this is as close as I could get to reproducing the error)

### CODE REVIEW

File: lib/zxcvbn/tester.rb [hunk 1]
```diff
@@ -21,15 +23,21 @@ def initialize
     end
 
     def test(password, user_inputs = [])
-      PasswordStrength.new(@data).test(password, user_inputs)
+      PasswordStrength.new(@data).test(password, sanitize(user_inputs))
```
Comment (johnsyweb) [inline lib/zxcvbn/tester.rb:26]: Adding some specs below revealed that non-string values can be added to the dictionary and they cause a similar runtime error to the one that caused me to open this gem.

File: spec/tester_spec.rb [hunk 2]
```diff
@@ -24,33 +27,73 @@
...
+      it "ignores those entries" do
+        expect(tester.test("themeforest")).to have_attributes(entropy: 0, score: 0, crack_time: 0)
+      end
+    end
   end
 
-  context 'nil password' do
+  context "nil password" do
     specify do
-      expect { tester.test(nil) }.to_not raise_error
```
Comment (johnsyweb) [inline spec/tester_spec.rb:53]: `expect { tester.test(nil) }.to_not raise_error` is one of my personal bugbears!

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+ - Invalid user dictionaries are handled more robustly [[#28][28]
```
Comment (jacobbednarz) [inline CHANGELOG.md:10]: Looks like an extra square brace in here

```suggestion
 - Invalid user dictionaries are handled more robustly [#28][28]
```

File: lib/zxcvbn/matchers/l33t.rb [hunk 4]
```diff
@@ -31,8 +31,8 @@ def matches(password)
               token = password[match.i..match.j]
               next if token.downcase == match.matched_word.downcase
               match_substitutions = {}
-              substitution.each do |substitution, letter|
-                match_substitutions[substitution] = letter if token.include?(substitution)
+              substitution.each do |s, letter|
```
Comment (jacobbednarz) [inline lib/zxcvbn/matchers/l33t.rb:34]: I was thinking `inject`/`collect` might be a better fit here but looking at it again; I'm not sure if it is more readable as is.

Comment (johnsyweb) [reply lib/zxcvbn/matchers/l33t.rb:34]: I wondered about refactoring this method entirely but decided against it for now.

File: CHANGELOG.md (hunk 3 above)
Comment (johnsyweb) [reply CHANGELOG.md]: I missed the closing bracket. Addressed in 21dd9f76576ef8ab61908826d6dc9639d658e8e5.

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+ - Invalid user dictionaries are handled more robustly [[#28]][28]
+
+[28]: https://github.com/envato/zxcvbn-ruby/pull/28
```
Comment (lucascaton) [inline CHANGELOG.md:12]: I think GitHub automatically adds a link to PRs whenever you add a # before its number.

```suggestion
 - Invalid user dictionaries are handled more robustly [#28]
```

File: spec/tester_spec.rb [hunk 6]
```diff
@@ -1,4 +1,7 @@
-require 'spec_helper'
+# coding: utf-8
```
Comment (lucascaton) [inline spec/tester_spec.rb:1]: Is this necessary?

Comment (johnsyweb) [reply spec/tester_spec.rb:1]: It's not. Will remove in a subsequent PR.

**Pull request merged**